README.md
# Psocake

Making Data Analysis for Free-Electron Lasers for a piece of cake

## Getting Started at LCLS

Serial Crystallography:
https://confluence.slac.stanford.edu/display/PSDM/Psocake+SFX+tutorial

Single Particle Imaging:
https://confluence.slac.stanford.edu/display/PSDM/Psocake+SPI+tutorial


doc/ChangeLog
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id$
#
# Description:
#  ChangeLog file for package psocake
#------------------------------------------------------------------------

Package author: Chunhong "Chuck" Yoon

Please describe any modifications that you made to the package in the
reverse time order.

Tag: V00-01-55
2017-2-27 Chunhong Yoon
- Disable logbook

Tag: V00-01-54
2017-1-12 Chunhong Yoon
- Deploy detector distance when manually deploy button is used

Tag: V00-01-53
2017-1-5 David Schneider
- conditional import on logbook to get psocake to run against conda
	
Tag: V00-01-52
2016-12-18 Chunhong Yoon
- Fix PeakFinder bug by removing set_mask function from PyAlgos

Tag: V00-01-51
2016-12-15 Chunhong Yoon
- Fix bug in peak finding for 2D detectors

Tag: V00-01-50
2016-12-09 Chunhong Yoon
- Psocake lite contains peak finder panel

Tag: V00-01-49
2016-11-21 Chunhong Yoon
- Better handling of epic variables, such as photon energy
- Correct handling of Rayonix geometry from psgeom

Tag: V00-01-48
2016-10-26 Chunhong Yoon
- Fix cxi file generation
- Update indexing in small data panel
- Set auto minimum image level to zero

Tag: V00-01-47
2016-10-19 Chunhong Yoon
- Fix CrystFEL geometry to psana geometry

Tag: V00-01-46
2016-10-04 Chunhong Yoon
- clean up for User's meeting

Tag: V00-01-45
2016-08-25 Chunhong Yoon
- Add label propagation

Tag: V00-01-44
2016-08-21 Chunhong Yoon
- Rayonix indexing
- Deploy maunally centred geometry for Rayonix

Tag: V00-01-43
2016-08-19 Chunhong Yoon
- Add crystfel geom to psana geometry

Tag: V00-01-42
2016-08-18 Chunhong Yoon
- Remove plot of polygon mask

Tag: V00-01-41
2016-08-18 Chunhong Yoon
- Fix clenEpics name generated from source

Tag: V00-01-40
2016-08-14 Chunhong Yoon
- Remove hardlink for psgeom
- Fix polygon mask in toggle mode

Tag: V00-01-39
2016-08-13 Chunhong Yoon
- Fix detector distance update
- Auto update small data name
- Peak finding batch job with median or radial background subtraction

Tag: V00-01-38
2016-08-13 Chunhong Yoon
- allow interactive median and radial background subtraction

Tag: V00-01-37
2016-08-05 Chunhong Yoon
- fix divide by zero when photon energy is zero
- change hit finder

Tag: V00-01-36
2016-08-01 Chunhong Yoon
- fix hitfinder bug with mask names

Tag: V00-01-35
2016-08-01 Chunhong Yoon
- fix bugs during refactoring

Tag: V00-01-34
2016-07-31 Chunhong Yoon
- refactor all panels

Tag: V00-01-33
2016-07-29 Chunhong Yoon
- added image panel

Tag: V00-01-32
2016-07-29 Chunhong Yoon
- added user feedback

Tag: V00-01-31
2016-07-26 Chunhong Yoon
- Majore refactoring of code

Tag: V00-01-30
2016-07-22 Chunhong Yoon
- Deploy geometry based on cirlce ROI

Tag: V00-01-29
2016-07-22 Chunhong Yoon
- Launcher for all threads

Tag: V00-01-28
2016-07-22 Chunhong Yoon
- Init logger

Tag: V00-01-27
2016-07-22 Chunhong Yoon
- Fix common mode correction

Tag: V00-01-26
2016-07-18 Chunhong Yoon
- changed nm to angstrom for unitcell display
- added detector centre indicator

Tag: V00-01-25
2016-07-14 Chunhong Yoon
- Added label panel
- Decouple elog with gui

Tag: V00-01-24
2016-07-05 Chunhong Yoon
- Masking for 2D detectors

Tag: V00-01-23
2016-06-29 Chunhong Yoon
- fix LitPixels

Tag: V00-01-22
2016-06-29 Chunhong Yoon
- Peak finding on Opal

Tag: V00-01-21
2016-06-29 Chunhong Yoon
- Handle area detectors with 2D shape
- Auto level for various image properties

Tag: V00-01-20
2016-06-17 Chunhong Yoon
- Added hybrid gain for cspad types

Tag: V00-01-19
2016-06-15 Chunhong Yoon
- Bug fix: use clen only at CXI

Tag: V00-01-18
2016-06-06 Chunhong Yoon
- Demo version

Tag: V00-01-17
2016-05-23 Chunhong Yoon
- Allow detector distance scan
- Triple integration ring

Tag: V00-01-16
2016-05-23 Chunhong Yoon
- Add indexing result display

Tag: V00-01-15
2016-05-22 Chunhong Yoon
- Add CrystFEL indexing panel

Tag: V00-01-14
2016-05-18 Chunhong Yoon
- Fix 2048 peak finder limit

Tag: V00-01-13
2016-05-13 Chunhong Yoon
- Removed peak finder v2

Tag: V00-01-12
2016-05-12 Christopher O'Grady
- For cspad2x2 user masks, save both DAQ/natural shape arrays in npy files

Tag: V00-01-11
2016-05-08 Chunhong Yoon
- Add back generatePowder

Tag: V00-01-10
2016-05-08 Chunhong Yoon
- Fix pedestal display
- Add common mode display

Tag: V00-01-09
2016-05-02 Chunhong Yoon
- Better detector information display

Tag: V00-01-08

src/gui.py
# GUI for browsing LCLS area detectors. Tune hit finding parameters and common mode correction.

# TODO: Multiple subplots or grid of images
# TODO: dropdown menu for available detectors, dropdown menu for small data
# TODO: When front and back detectors given, display both
# TODO: Display acqiris
# TODO: Downsampler
# TODO: Run from ffb
# TODO: script for histogram of lit pixels, percentile display

import sys, signal
import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtCore, QtGui
import pyqtgraph.console
from pyqtgraph.dockarea import *
from pyqtgraph.dockarea.Dock import DockLabel
from pyqtgraph.parametertree import Parameter, ParameterTree#, ParameterItem, registerParameterType
import psana
import h5py
import argparse
import time
import subprocess
import os.path

# Panel modules
import diffractionGeometryPanel, crystalIndexingPanel, SmallDataPanel, ExperimentPanel
import PeakFindingPanel, HitFinderPanel, MaskPanel, LabelPanel, ImagePanel, RoiPanel
import ImageControlPanel, MousePanel, ImageStackPanel
import _colorScheme as color
from _version import __version__

parser = argparse.ArgumentParser()
parser.add_argument('expRun', nargs='?', default=None, help="Psana-style experiment/run string in the format (e.g. exp=cxi06216:run=22). This option trumps -e and -r options.")
parser.add_argument("-e","--exp", help="Experiment name (e.g. cxis0813 ). This option is ignored if expRun option is used.", default="", type=str)
parser.add_argument("-r","--run", help="Run number. This option is ignored if expRun option is used.",default=0, type=int)
parser.add_argument("-d","--det", help="Detector alias or DAQ name (e.g. DscCsPad or CxiDs1.0:Cspad.0), default=''",default="", type=str)
parser.add_argument("-n","--evt", help="Event number (e.g. 1), default=0",default=0, type=int)
parser.add_argument("--localCalib", help="Use local calib directory. A calib directory must exist in your current working directory.", action='store_true')
parser.add_argument("-o","--outDir", help="Use this directory for output instead.", default=None, type=str)
parser.add_argument("-v", help="verbosity level, default=0",default=0, type=int)
parser.add_argument('--version', action='version',
                    version='%(prog)s {version}'.format(version=__version__))
parser.add_argument("-m","--mode", help="Mode sets the combination of panels available on the GUI, options: {lite,sfx,spi,all}",default="lite", type=str)
args = parser.parse_args()

class Window(QtGui.QMainWindow):
    global ex

    def previewEvent(self, eventNumber):
        ex.eventNumber = eventNumber
        ex.calib, ex.data = ex.img.getDetImage(ex.eventNumber)
        ex.img.w1.setImage(ex.data, autoRange=False, autoLevels=False, autoHistogramRange=False)
        ex.exp.p.param(ex.exp.exp_grp, ex.exp.exp_evt_str).setValue(ex.eventNumber)

    def keyPressEvent(self, event):
        super(Window, self).keyPressEvent(event)
        if args.mode == "all":
            if type(event) == QtGui.QKeyEvent:
                path = ["", ""]
                if event.key() == QtCore.Qt.Key_1:
                    path[1] = "Single"
                    data = True
                    if ex.evtLabels.labelA == True : data = False
                    ex.evtLabels.paramUpdate(path, data)
                elif event.key() == QtCore.Qt.Key_2:
                    path[1] = "Multi"
                    data = True
                    if ex.evtLabels.labelB == True : data = False
                    ex.evtLabels.paramUpdate(path, data)
                elif event.key() == QtCore.Qt.Key_3:
                    path[1] = "Dunno"
                    data = True
                    if ex.evtLabels.labelC == True : data = False
                    ex.evtLabels.paramUpdate(path, data)
                elif event.key() == QtCore.Qt.Key_Period:
                    if ex.small.w9.getPlotItem().listDataItems() != []:
                        idx = -1
                        array = np.where(ex.small.quantifierEvent >= ex.eventNumber)
                        if array[0].size != 0:
                            idx = array[0][0]
                            if ex.small.quantifierEvent[idx] == ex.eventNumber: idx += 1
                            if idx < (ex.small.quantifierEvent.size): self.previewEvent(ex.small.quantifierEvent[idx])
                elif event.key() == QtCore.Qt.Key_N:
                    if ex.eventNumber < (ex.exp.eventTotal - 1): self.previewEvent(ex.eventNumber+1)
                elif event.key() == QtCore.Qt.Key_Comma:
                    if ex.small.w9.getPlotItem().listDataItems() != []:
                        idx = -1
                        array = np.where(ex.small.quantifierEvent <= ex.eventNumber)
                        if array[0].size != 0:
                            idx = array[0][array[0].size - 1]
                            if ex.small.quantifierEvent[idx] == ex.eventNumber: idx -= 1
                            if ex.small.quantifierEvent[idx] != 0: self.previewEvent(ex.small.quantifierEvent[idx])
                elif event.key() == QtCore.Qt.Key_P:
                    if ex.eventNumber != 0: self.previewEvent(ex.eventNumber-1)
                ex.evtLabels.refresh()

class MainFrame(QtGui.QWidget):
    """
    The main frame of the application
    """        
    def __init__(self, arg_list):
        super(MainFrame, self).__init__()
        self.args = args

        # Set up tolerance
        self.eps = np.finfo("float64").eps
        self.firstUpdate = True
        self.operationModeChoices = ['none','masking']
        self.operationMode =  self.operationModeChoices[0] # Masking mode, Peak finding mode
        # Init experiment parameters from args
        if args.expRun is not None and ':run=' in args.expRun:
            self.experimentName = args.expRun.split('exp=')[-1].split(':')[0]
            self.runNumber = int(args.expRun.split('run=')[-1])
        else:
            self.experimentName = args.exp
            self.runNumber = int(args.run)
        self.detInfo = args.det
        self.detAlias = None
        self.eventNumber = int(args.evt)

        # Directories
        self.psocakeDir = None
        self.psocakeRunDir = None
        self.elogDir = None
        self.rootDir = None
        self.writeAccess = True

        # Init variables
        self.det = None
        self.detnames = None
        self.detInfoList = None
        self.isCspad = False
        self.evt = None
        self.eventID = ""
        self.hasExperimentName = False
        self.hasRunNumber = False
        self.hasDetInfo = False
        self.pixelIndAssem = None

        # Init diffraction geometry parameters
        self.coffset = 0.0
        self.clen = 0.0
        self.clenEpics = 0.0
        self.epics = None
        self.detectorDistance = 0.0
        self.photonEnergy = None
        self.wavelength = None
        self.pixelSize = None
        self.resolutionRingsOn = False
        self.resolution = None
        self.resolutionUnits = 0

        # Init variables
        self.calib = None # ndarray detector image
        self.data = None # assembled detector image
        self.cx = 0 # detector centre x
        self.cy = 0 # detector centre y

        ## Switch to using white background and black foreground
        pg.setConfigOption('background', color.grayLite_hex)
        pg.setConfigOption('foreground', color.darkRed_hex)

        ########################################
        # Instantiate panels
        ########################################
        self.exp = ExperimentPanel.ExperimentInfo(self)
        self.geom = diffractionGeometryPanel.DiffractionGeometry(self)
        self.index = crystalIndexingPanel.CrystalIndexing(self)
        self.small = SmallDataPanel.SmallData(self)
        self.evtLabels = LabelPanel.Labels(self)
        self.pk = PeakFindingPanel.PeakFinding(self)
        self.hf = HitFinderPanel.HitFinder(self)
        self.mk = MaskPanel.MaskMaker(self)
        self.img = ImagePanel.ImageViewer(self)
        self.roi = RoiPanel.RoiHistogram(self)
        self.control = ImageControlPanel.ImageControl(self)
        self.mouse = MousePanel.Mouse(self)
        self.stack = ImageStackPanel.ImageStack(self)

        self.initUI()

    def initUI(self):
        ## Define a top-level widget to hold everything
        self.win = Window()
        self.area = DockArea()
        self.win.setCentralWidget(self.area)
        self.win.resize(1300,650)
        self.win.setWindowTitle('psocake')

        # Set the color scheme
        def updateStylePatched(self):
            r = '3px'
            fg = color.cardinalRed_hex
            bg = color.sandstone100_rgb
            border = "k"

            if self.orientation == 'vertical':
                self.vStyle = """DockLabel {
                    background-color : %s;
                    color : %s;
                    border-top-right-radius: 0px;
                    border-top-left-radius: %s;
                    border-bottom-right-radius: 0px;
                    border-bottom-left-radius: %s;
                    border-top: 1px solid %s;
                    border-left: 1px solid %s;
                    border-right: 1px solid %s;
                    border-width: 0x;
                    border-right: 2px solid %s;
                    padding-top: 3px;
                    padding-bottom: 3px;
                    font-size: 18px;
                }""" % (bg, fg, r, r, border, fg, fg, fg)
                self.setStyleSheet(self.vStyle)
            else:
                self.hStyle = """DockLabel {
                    background-color : %s;
                    color : %s;
                    border-top-right-radius: %s;
                    border-top-left-radius: %s;
                    border-bottom-right-radius: 0px;
                    border-bottom-left-radius: 0px;
                    border-width: 0px;
                    border-bottom: 2px solid %s;
                    border-top: 1px solid %s;
                    border-left: 1px solid %s;
                    border-right: 1px solid %s;
                    padding-left: 13px;
                    padding-right: 13px;
                    font-size: 18px
                }""" % (bg, fg, r, r, border, fg, fg, fg)
                self.setStyleSheet(self.hStyle)
        DockLabel.updateStyle = updateStylePatched

        if args.mode == 'sfx':
            # Dock positions on the main frame
            self.area.addDock(self.mouse.d5, 'left')  ## place d5 at left edge of d1
            self.area.addDock(self.control.d6, 'bottom', self.mouse.d5)    ## place d1 at left edge of dock area
            self.area.addDock(self.stack.d7, 'bottom', self.mouse.d5)
            self.area.addDock(self.img.d1, 'bottom', self.mouse.d5)    ## place d1 at left edge of dock area
            self.area.moveDock(self.img.d1, 'above', self.stack.d7)

            self.area.addDock(self.exp.d2, 'right')     ## place d2 at right edge of dock area
            self.area.addDock(self.pk.d9, 'bottom', self.exp.d2)
            self.area.addDock(self.mk.d12, 'bottom',self.exp.d2)
            self.area.addDock(self.index.d14, 'bottom', self.exp.d2)
            self.area.moveDock(self.pk.d9, 'above', self.mk.d12)
            self.area.moveDock(self.index.d14, 'above', self.mk.d12)

            self.area.addDock(self.geom.d3, 'bottom', self.exp.d2)    ## place d3 at bottom edge of d1
            self.area.addDock(self.roi.d4, 'bottom', self.exp.d2)    ## place d4 at right edge of dock area
            self.area.moveDock(self.geom.d3, 'above', self.exp.d2)
            self.area.moveDock(self.roi.d4, 'above', self.exp.d2)

            self.area.addDock(self.small.dSmall, 'right')#, self.exp.d2)
            self.area.moveDock(self.exp.d2, 'above', self.geom.d3)
        elif args.mode == 'spi':
            # Dock positions on the main frame
            self.area.addDock(self.mouse.d5, 'left')  ## place d5 at left edge of d1
            self.area.addDock(self.control.d6, 'bottom', self.mouse.d5)    ## place d1 at left edge of dock area
            self.area.addDock(self.stack.d7, 'bottom', self.mouse.d5)
            self.area.addDock(self.img.d1, 'bottom', self.mouse.d5)    ## place d1 at left edge of dock area
            self.area.moveDock(self.img.d1, 'above', self.stack.d7)

            self.area.addDock(self.exp.d2, 'right')     ## place d2 at right edge of dock area
            self.area.addDock(self.mk.d12, 'bottom',self.exp.d2)
            self.area.addDock(self.hf.d13, 'bottom', self.exp.d2)
            self.area.moveDock(self.hf.d13, 'above', self.mk.d12)

            self.area.addDock(self.geom.d3, 'bottom', self.exp.d2)    ## place d3 at bottom edge of d1
            self.area.addDock(self.roi.d4, 'bottom', self.exp.d2)    ## place d4 at right edge of dock area
            self.area.moveDock(self.geom.d3, 'above', self.exp.d2)
            self.area.moveDock(self.roi.d4, 'above', self.exp.d2)

            self.area.addDock(self.small.dSmall, 'right')#, self.exp.d2)
            self.area.moveDock(self.exp.d2, 'above', self.geom.d3)
        elif args.mode == 'all':
            # Dock positions on the main frame
            self.area.addDock(self.mouse.d5, 'left')  ## place d5 at left edge of d1
            self.area.addDock(self.control.d6, 'bottom', self.mouse.d5)  ## place d1 at left edge of dock area
            self.area.addDock(self.stack.d7, 'bottom', self.mouse.d5)
            self.area.addDock(self.img.d1, 'bottom', self.mouse.d5)  ## place d1 at left edge of dock area
            self.area.moveDock(self.img.d1, 'above', self.stack.d7)

            self.area.addDock(self.exp.d2, 'right')  ## place d2 at right edge of dock area
            self.area.addDock(self.pk.d9, 'bottom', self.exp.d2)
            self.area.addDock(self.mk.d12, 'bottom', self.exp.d2)
            self.area.addDock(self.hf.d13, 'bottom', self.exp.d2)
            self.area.addDock(self.index.d14, 'bottom', self.exp.d2)
            self.area.moveDock(self.pk.d9, 'above', self.mk.d12)
            self.area.moveDock(self.hf.d13, 'above', self.mk.d12)
            self.area.moveDock(self.index.d14, 'above', self.mk.d12)

            self.area.addDock(self.geom.d3, 'bottom', self.exp.d2)  ## place d3 at bottom edge of d1
            self.area.addDock(self.roi.d4, 'bottom', self.exp.d2)  ## place d4 at right edge of dock area
            self.area.moveDock(self.geom.d3, 'above', self.exp.d2)
            self.area.moveDock(self.roi.d4, 'above', self.exp.d2)

            self.area.addDock(self.small.dSmall, 'right')  # , self.exp.d2)
            self.area.moveDock(self.exp.d2, 'above', self.geom.d3)

            self.area.addDock(self.evtLabels.dLabels, 'bottom', self.small.dSmall)
        else: # lite
            # Dock positions on the main frame
            self.area.addDock(self.mouse.d5, 'left')  ## place d5 at left edge of d1
            self.area.addDock(self.control.d6, 'bottom', self.mouse.d5)  ## place d1 at left edge of dock area
            self.area.addDock(self.stack.d7, 'bottom', self.mouse.d5)
            self.area.addDock(self.img.d1, 'bottom', self.mouse.d5)  ## place d1 at left edge of dock area
            self.area.moveDock(self.img.d1, 'above', self.stack.d7)

            self.area.addDock(self.roi.d4, 'right')
            self.area.addDock(self.exp.d2, 'above', self.roi.d4)
            self.area.addDock(self.mk.d12, 'bottom', self.exp.d2)
            self.area.addDock(self.pk.d9, 'above', self.mk.d12)

        ###############
        ### Threads ###
        ###############
        # Making powder patterns
        self.thread = []
        self.threadCounter = 0

        # Setup input parameters
        if self.experimentName is not "":
            self.hasExperimentName = True
            self.exp.p.param(self.exp.exp_grp, self.exp.exp_name_str).setValue(self.experimentName)
            self.exp.updateExpName(self.experimentName)
        if self.runNumber is not 0:
            self.hasRunNumber = True
            self.exp.p.param(self.exp.exp_grp, self.exp.exp_run_str).setValue(self.runNumber)
            self.exp.updateRunNumber(self.runNumber)
        if self.detInfo is not "":
            self.hasDetInfo = True
            self.exp.p.param(self.exp.exp_grp, self.exp.exp_detInfo_str).setValue(self.detInfo)
            self.exp.updateDetInfo(self.detInfo)
        self.exp.p.param(self.exp.exp_grp, self.exp.exp_evt_str).setValue(self.eventNumber)
        self.exp.updateEventNumber(self.eventNumber)

        # Indicate centre of detector
        self.geom.drawCentre()

        # Try mouse over crosshair
        self.xhair = self.img.w1.getView()
        self.vLine = pg.InfiniteLine(angle=90, movable=False)
        self.hLine = pg.InfiniteLine(angle=0, movable=False)
        self.xhair.addItem(self.vLine, ignoreBounds=True)
        self.xhair.addItem(self.hLine, ignoreBounds=True)
        self.vb = self.xhair.vb
        self.label = pg.LabelItem()
        self.mouse.w5.addItem(self.label)

        def mouseMoved(evt):
            pos = evt[0]  ## using signal proxy turns original arguments into a tuple
            if self.xhair.sceneBoundingRect().contains(pos):
                mousePoint = self.vb.mapSceneToView(pos)
                indexX = int(mousePoint.x())
                indexY = int(mousePoint.y())

                # update crosshair position
                self.vLine.setPos(mousePoint.x())
                self.hLine.setPos(mousePoint.y())
                # get pixel value, if data exists
                if self.data is not None:
                    if indexX >= 0 and indexX < self.data.shape[0] \
                       and indexY >= 0 and indexY < self.data.shape[1]:
                        if self.mk.maskingMode > 0:
                            modeInfo = self.mk.masking_mode_message
                        else:
                            modeInfo = ""
                        pixelInfo = "<span style='color: " + color.cardinalRed_hex + "; font-size: 24pt;'>x=%0.1f y=%0.1f I=%0.1f </span>"
                        self.label.setText(modeInfo + pixelInfo % (mousePoint.x(), mousePoint.y(), self.data[indexX,indexY]))

        def mouseClicked(evt):
            mousePoint = self.vb.mapSceneToView(evt[0].scenePos())
            indexX = int(mousePoint.x())
            indexY = int(mousePoint.y())

            if self.data is not None:
                # Mouse click
                if indexX >= 0 and indexX < self.data.shape[0] \
                    and indexY >= 0 and indexY < self.data.shape[1]:
                    print "mouse clicked: ", mousePoint.x(), mousePoint.y(), self.data[indexX,indexY]
                    if self.mk.maskingMode > 0:
                        self.initMask()
                        if self.mk.maskingMode == 1:
                            # masking mode
                            self.mk.userMaskAssem[indexX,indexY] = 0
                        elif self.mk.maskingMode == 2:
                            # unmasking mode
                            self.mk.userMaskAssem[indexX,indexY] = 1
                        elif self.mk.maskingMode == 3:
                            # toggle mode
                            self.mk.userMaskAssem[indexX,indexY] = (1-self.mk.userMaskAssem[indexX,indexY])
                        self.displayMask()

                        self.mk.userMask = self.det.ndarray_from_image(self.evt,self.mk.userMaskAssem, pix_scale_size_um=None, xy0_off_pix=None)
                        self.algInitDone = False
                        self.parent.pk.updateClassification()

        # Signal proxy
        self.proxy_move = pg.SignalProxy(self.xhair.scene().sigMouseMoved, rateLimit=30, slot=mouseMoved)
        self.proxy_click = pg.SignalProxy(self.xhair.scene().sigMouseClicked, slot=mouseClicked)
        self.win.show()

    # If anything changes in the parameter tree, print a message
    def change(self, panel, changes):
        for param, change, data in changes:
            path = panel.childPath(param)
            if args.v >= 1:
                print('  path: %s'% path)
                print('  change:    %s'% change)
                print('  data:      %s'% str(data))
                print('  ----------')
            self.update(path,change,data)

    def update(self, path, change, data):
        if args.v >= 1: print "path: ", path
        ################################################
        # experiment parameters
        ################################################
        if path[0] == self.exp.exp_grp or path[0] == self.exp.disp_grp:
            self.exp.paramUpdate(path, change, data)
        ################################################
        # peak finder parameters
        ################################################
        if path[0] == self.pk.hitParam_grp:
            self.pk.paramUpdate(path, change, data)
        ################################################
        # hit finder parameters
        ################################################
        if path[0] == self.hf.spiParam_grp:
            self.hf.paramUpdate(path, change, data)
        ################################################
        # diffraction geometry parameters
        ################################################
        if path[0] == self.geom.geom_grp:
            self.geom.paramUpdate(path, change, data)
        ################################################
        # quantifier parameters
        ################################################
        if path[0] == self.small.quantifier_grp:
            self.small.paramUpdate(path, change, data)
        ################################################
        # masking parameters
        ################################################
        if path[0] == self.mk.mask_grp or path[0] == self.mk.powder_grp:
            self.mk.paramUpdate(path, change, data)
        ################################################
        # crystal indexing parameters
        ################################################
        if path[0] == self.index.index_grp or path[0] == self.index.launch_grp:
            self.index.paramUpdate(path, change, data)
        ################################################
        # label parameters
        ################################################
        if path[0] == self.evtLabels.labels_grp:
            self.evtLabels.paramUpdate(path, change, data)
            self.evtLabels.refresh()

def main():
    global ex
    signal.signal(signal.SIGINT, signal.SIG_DFL)
    app = QtGui.QApplication(sys.argv)
    ex = MainFrame(sys.argv)
    sys.exit(app.exec_())

## Start Qt event loop unless running in interactive mode or using pyside.
if __name__ == '__main__':
    main()
